device/common/include/traccc/device/impl/mutex.ipp

```diff
@@ -0,0 +1,48 @@
+/**
+ * traccc library, part of the ACTS project (R&D line)
+ *
+ * (c) 2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+#include <cassert>
+#include <vecmem/memory/device_atomic_ref.hpp>
+
+namespace traccc::device {
+template <typename T>
+mutex<T>::mutex(T* p) : m_atomic(*p) {}
```

Comment:
Also: Since we're de-referencing it right away (i.e. a `nullptr` is never an option), the interface should require a reference instead of a pointer. Putting the burden on the caller for not providing an invalid pointer.

Reply:
Sure. :+1: